packages/@aws-cdk/core/lib/stack.ts

```diff
@@ -231,10 +238,16 @@ export class Stack extends Construct implements ITaggable {
     this.tags = new TagManager(TagType.KEY_VALUE, 'aws:cdk:stack', props.tags);
 
     if (!VALID_STACK_NAME_REGEX.test(this.stackName)) {
-      throw new Error(`Stack name must match the regular expression: ${VALID_STACK_NAME_REGEX.toString()}, got '${name}'`);
+      throw new Error(`Stack name must match the regular expression: ${VALID_STACK_NAME_REGEX.toString()}, got '${id}'`);
     }
 
-    this.templateFile = `${this.stackName}.template.json`;
+    this.artifactId = this.calculateStackName();
```

Comment:
I don't like this line. `calculateStackName` is now a misnomer for what it's being used for. Maybe rename to something like `uniqueStackName()` or something? Or `treeBasedStackName()`?

Reply:
Yes, good point. I'll rename

Reply:
renamed to `generateUniqueStackName`